File: problems/p0159_longest_substring_with_at_most_two_distinct_characters/typescript/_solution.test.ts
import { describe, test, expect } from "vitest";
import { solutions } from "./solution";

describe.each(solutions)("lengthOfLongestSubstringTwoDistinct - %o", (lengthOfLongestSubstringTwoDistinct) => {
  // --- Provided examples ---

  test("example 1 - eceba", () => {
    expect(lengthOfLongestSubstringTwoDistinct("eceba")).toBe(3);
  });

  test("example 2 - ccaabbb", () => {
    expect(lengthOfLongestSubstringTwoDistinct("ccaabbb")).toBe(5);
  });

  // --- Edge cases ---

  test("single character", () => {
    expect(lengthOfLongestSubstringTwoDistinct("a")).toBe(1);
  });

  test("two distinct characters - entire string", () => {
    expect(lengthOfLongestSubstringTwoDistinct("aabb")).toBe(4);
  });

  test("all same characters", () => {
    expect(lengthOfLongestSubstringTwoDistinct("aaaaa")).toBe(5);
  });

  test("two characters only", () => {
    expect(lengthOfLongestSubstringTwoDistinct("ab")).toBe(2);
  });

  test("empty string", () => {
    expect(lengthOfLongestSubstringTwoDistinct("")).toBe(0);
  });

  // --- Longer strings ---

  test("longest substring at end", () => {
    expect(lengthOfLongestSubstringTwoDistinct("abcbbbb")).toBe(4);
  });

  test("longest substring at start", () => {
    expect(lengthOfLongestSubstringTwoDistinct("aaaabcd")).toBe(4);
  });

  test("alternating characters", () => {
    expect(lengthOfLongestSubstringTwoDistinct("ababab")).toBe(6);
  });

  test("three distinct - middle is longest", () => {
    expect(lengthOfLongestSubstringTwoDistinct("abbbca")).toBe(4);
  });

  // --- Boundary: exactly two distinct in entire string ---

  test("exactly two distinct - entire string qualifies", () => {
    expect(lengthOfLongestSubstringTwoDistinct("aababba")).toBe(7);
  });

  // --- Many distinct characters ---

  test("all distinct characters", () => {
    expect(lengthOfLongestSubstringTwoDistinct("abcdef")).toBe(2);
  });
});

File: problems/p0159_longest_substring_with_at_most_two_distinct_characters/java/p0159_longest_substring_with_at_most_two_distinct_characters/_SolutionTest.java
package p0159_longest_substring_with_at_most_two_distinct_characters;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class _SolutionTest {

    static Stream<Solution> solutions() {
        return Stream.of(
                new Default(),
                new WithStreams()
        );
    }

    // --- Provided examples ---

    @ParameterizedTest
    @MethodSource("solutions")
    void example1_eceba(Solution solution) {
        assertEquals(3, solution.lengthOfLongestSubstringTwoDistinct("eceba"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void example2_ccaabbb(Solution solution) {
        assertEquals(5, solution.lengthOfLongestSubstringTwoDistinct("ccaabbb"));
    }

    // --- Edge cases ---

    @ParameterizedTest
    @MethodSource("solutions")
    void singleCharacter(Solution solution) {
        assertEquals(1, solution.lengthOfLongestSubstringTwoDistinct("a"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void twoDistinctCharacters_entireString(Solution solution) {
        assertEquals(4, solution.lengthOfLongestSubstringTwoDistinct("aabb"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void allSameCharacters(Solution solution) {
        assertEquals(5, solution.lengthOfLongestSubstringTwoDistinct("aaaaa"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void twoCharactersOnly(Solution solution) {
        assertEquals(2, solution.lengthOfLongestSubstringTwoDistinct("ab"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void emptyString(Solution solution) {
        assertEquals(0, solution.lengthOfLongestSubstringTwoDistinct(""));
    }

    // --- Longer strings ---

    @ParameterizedTest
    @MethodSource("solutions")
    void longestSubstringAtEnd(Solution solution) {
        assertEquals(4, solution.lengthOfLongestSubstringTwoDistinct("abcbbbb"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void longestSubstringAtStart(Solution solution) {
        assertEquals(4, solution.lengthOfLongestSubstringTwoDistinct("aaaabcd"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void alternatingCharacters(Solution solution) {
        assertEquals(6, solution.lengthOfLongestSubstringTwoDistinct("ababab"));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void threeDistinctCharacters_middleIsLongest(Solution solution) {
        assertEquals(4, solution.lengthOfLongestSubstringTwoDistinct("abbbca"));
    }

    // --- Boundary: exactly two distinct in entire string ---

    @ParameterizedTest
    @MethodSource("solutions")
    void exactlyTwoDistinct_entireStringQualifies(Solution solution) {
        assertEquals(7, solution.lengthOfLongestSubstringTwoDistinct("aababba"));
    }

    // --- Many distinct characters ---

    @ParameterizedTest
    @MethodSource("solutions")
    void allDistinctCharacters(Solution solution) {
        assertEquals(2, solution.lengthOfLongestSubstringTwoDistinct("abcdef"));
    }
}
